### DISCUSSION TITLE
[Idea] OP Stack - Deploy Config Generator

### DISCUSSION THREAD

**Discussion by gosunuts** [ethereum-optimism/developers#559] [Ideas]
### Idea

During the deployment of the OP stack, one of the most challenging aspects was managing the deployment configs.
It's difficult to find all the configuration specifications and use cases. The only complete reference is the [golang code](https://github.com/ethereum-optimism/optimism/blob/develop/op-chain-ops/genesis/config.go), but It is hard to find for chain operators.

Another issue is the format of the deployment configuration. JSON, being unordered and uncategorized, cannot effectively represent empty values. People have trouble distinguishing between numerous fields that need to be changed, and need to be kept.

As an alternative, I suggest generating the deployment config using [jsonnet](https://github.com/google/go-jsonnet). it offers several advantages:

- It can be categorized same as golang configs.
- It can handle three types of missing configurations:
  - `Required`: Throws an error if missing.
  - `Default`: Uses a default value if missing.
  - `Optional`: Returns an empty if missing.
- This approach enables the specification of all configurations.

It can be used in simple scripts ( shell script, python, golang ), and also used in interactive UIs.

For reference, here are some prototypes:

Jsonnet example: [jsonnet template](https://github.com/rabbitprincess/builderism/blob/main/init/3_config.jsonnet)
config generation example: [shell script for json generation](https://github.com/rabbitprincess/builderism/blob/main/init/3_generate_config.sh)

It is currently part of [builderism](https://github.com/rabbitprincess/builderism). If possible, I would like to see it adopted as a specification in the optimism tutorials.

**Comment by sbvegan** [comment]
Hey @rabbitprincess, I think this is a great idea. There is some efforts being put forward on the protocol front that will help ease this process, but in the meantime, a tool like this would be a great resource. The work around the [OP Stack Manager](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/op-stack-manager-arch.md) will help modularize the deployment configurations to make them easier to manage when deploying a new chain.
